File: server/internal/knowledge/structure.go
package knowledge

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"strings"
	"time"
	"unicode/utf8"
)

// 结构协议独立于旧 CLI 的严格 DTO；旧概念和评估继续引用原修订。
type ConceptSource struct {
	CollectionID string `json:"collection_id"`
	RevisionID   string `json:"revision_id"`
	DocumentID   string `json:"document_id"`
	NodeID       string `json:"node_id"`
	Quote        string `json:"quote"`
}
type ConceptClaim struct {
	Text       string `json:"text"`
	Conditions string `json:"conditions"`
	Sources    []int  `json:"sources"`
	Gap        string `json:"gap"`
}
type ConceptRelation struct {
	TargetID  string `json:"target_id"`
	Kind      string `json:"kind"`
	Suggested bool   `json:"suggested"`
	Sources   []int  `json:"sources"`
}
type ConceptContent struct {
	Description  string            `json:"description"`
	SourceStatus string            `json:"source_status"`
	Suggested    bool              `json:"suggested"`
	Sources      []ConceptSource   `json:"sources"`
	Claims       []ConceptClaim    `json:"claims"`
	Relations    []ConceptRelation `json:"relations"`
	ReplacedBy   []string          `json:"replaced_by"`
}
type StructureNode struct {
	ConceptRevision
	GoalID        string         `json:"goal_id"`
	Content       ConceptContent `json:"content"`
	LearningState string         `json:"learning_state"`
}
type StructureQuery struct {
	GoalID string
	RootID string
	Search string
	Cursor string
	Limit  int
}
type StructureEdge struct {
	SourceID string `json:"source_id"`
	ConceptRelation
}
type StructurePage struct {
	Version    int64           `json:"version"`
	Generation int64           `json:"generation"`
	Items      []StructureNode `json:"items"`
	Edges      []StructureEdge `json:"edges"`
	NextCursor string          `json:"next_cursor"`
	Partial    bool            `json:"partial"`
	Notice     string          `json:"notice"`
}
type StructureEdit struct {
	ConceptID string         `json:"concept_id"`
	GoalID    string         `json:"goal_id"`
	Name      string         `json:"name"`
	Content   ConceptContent `json:"content"`
}
type StructureCommand struct {
	OperationID   string          `json:"operation_id"`
	BaseVersion   int64           `json:"base_version"`
	Generation    int64           `json:"generation"`
	Kind          string          `json:"kind"`
	Reason        string          `json:"reason"`
	Edits         []StructureEdit `json:"edits"`
	Compensates   string          `json:"compensates,omitempty"`
	ActorDeviceID string          `json:"-"`
}
type StructureImpact struct {
	GoalIDs     []string `json:"goal_ids"`
	Contexts    int64    `json:"contexts"`
	Activities  int64    `json:"activities"`
	Contents    int64    `json:"contents"`
	Evidence    int64    `json:"evidence"`
	Fingerprint string   `json:"fingerprint"`
}
type StructureProposal struct {
	ID             string          `json:"id"`
	OperationID    string          `json:"operation_id"`
	Status         string          `json:"status"`
	Kind           string          `json:"kind"`
	Reason         string          `json:"reason"`
	BaseVersion    int64           `json:"base_version"`
	Generation     int64           `json:"generation"`
	Hash           string          `json:"hash"`
	Before         []StructureNode `json:"before"`
	After          []StructureNode `json:"after"`
	Impact         StructureImpact `json:"impact"`
	CurrentImpact  StructureImpact `json:"current_impact"`
	Compensates    string          `json:"compensates"`
	AppliedVersion int64           `json:"applied_version"`
	CreatedAt      time.Time       `json:"created_at"`
	DecisionReason string          `json:"decision_reason"`
	Replayed       bool            `json:"replayed"`
}
type StructureDecision struct {
	OperationID   string `json:"operation_id"`
	Hash          string `json:"hash"`
	Decision      string `json:"decision"`
	Reason        string `json:"reason"`
	ActorDeviceID string `json:"-"`
}
type StructureProposalPage struct {
	Items      []StructureProposal `json:"items"`
	NextCursor string              `json:"next_cursor"`
}
type StructureStore interface {
	ReadStructure(context.Context, StructureQuery) (StructurePage, error)
	ReadConcept(context.Context, string, string) (StructureNode, error)
	CreateStructureProposal(context.Context, StructureCommand) (StructureProposal, error)
	ReadStructureProposal(context.Context, string) (StructureProposal, error)
	ListStructureProposals(context.Context, string, int) (StructureProposalPage, error)
	DecideStructureProposal(context.Context, string, StructureDecision) (StructureProposal, error)
}

func StructureDigest(v any) string {
	raw, _ := json.Marshal(v)
	sum := sha256.Sum256(raw)
	return hex.EncodeToString(sum[:])
}
func structureText(v string, max int, required bool) bool {
	return utf8.ValidString(v) && len(v) <= max && !strings.ContainsRune(v, 0) && (!required || strings.TrimSpace(v) != "")
}
func (c StructureCommand) Validate() error {
	invalid := &Error{Code: CodeInvalidRequest}
	if !validUUID(c.OperationID) || !validUUID(c.ActorDeviceID) || c.BaseVersion < 0 || c.Generation < 1 || !structureText(c.Reason, 2000, true) {
		return invalid
	}
	if c.Kind == "compensate" {
		if !validUUID(c.Compensates) || len(c.Edits) != 0 {
			return invalid
		}
		return nil
	}
	if c.Compensates != "" || len(c.Edits) < 1 || len(c.Edits) > 20 {
		return invalid
	}
	if c.Kind != "edit" && c.Kind != "merge" && c.Kind != "split" {
		return invalid
	}
	seen := map[string]bool{}
	for _, e := range c.Edits {
		if !validUUID(e.ConceptID) || seen[e.ConceptID] || (e.GoalID != "" && !validUUID(e.GoalID)) || !structureText(e.Name, 300, true) {
			return invalid
		}
		seen[e.ConceptID] = true
		v := e.Content
		if !structureText(v.Description, 8000, false) || len(v.Sources) > 16 || len(v.Claims) > 16 || len(v.Relations) > 40 || len(v.ReplacedBy) > 20 {
			return invalid
		}
		switch v.SourceStatus {
		case "candidate", "included", "unverified", "conflict", "superseded":
		default:
			return invalid
		}
		if (v.SourceStatus == "superseded") != (len(v.ReplacedBy) > 0) {
			return invalid
		}
		for _, source := range v.Sources {
			if !validUUID(source.CollectionID) || !validUUID(source.RevisionID) || !validUUID(source.DocumentID) || !validUUID(source.NodeID) || !structureText(source.Quote, 2000, true) {
				return invalid
			}
		}
		indexes := func(ids []int) bool {
			if len(ids) > 16 {
				return false
			}
			for _, n := range ids {
				if n < 0 || n >= len(v.Sources) {
					return false
				}
			}
			return true
		}
		for _, claim := range v.Claims {
			if !structureText(claim.Text, 2000, true) || !structureText(claim.Conditions, 2000, false) || !structureText(claim.Gap, 2000, false) || !indexes(claim.Sources) || len(claim.Sources) == 0 && claim.Gap == "" {
				return invalid
			}
		}
		if v.SourceStatus == "conflict" && len(v.Claims) < 2 {
			return invalid
		}
		for _, edge := range v.Relations {
			if !validUUID(edge.TargetID) || edge.TargetID == e.ConceptID || !indexes(edge.Sources) {
				return invalid
			}
			switch edge.Kind {
			case "prerequisite", "related", "contrast", "part_of":
			default:
				return invalid
			}
		}
		for _, id := range v.ReplacedBy {
			if !validUUID(id) || id == e.ConceptID {
				return invalid
			}
		}
	}
	return nil
}

func (s *Service) SupportsStructure() bool { _, ok := s.store.(StructureStore); return ok }
func (s *Service) structureStore() (StructureStore, error) {
	v, ok := s.store.(StructureStore)
	if !ok {
		return nil, &Error{Code: CodeNotFound}
	}
	return v, nil
}
func (s *Service) ReadStructure(ctx context.Context, q StructureQuery) (StructurePage, error) {
	if q.Limit == 0 {
		q.Limit = 30
	}
	if q.Limit < 1 || q.Limit > 100 || !structureText(q.Search, 200, false) || q.GoalID != "" && !validUUID(q.GoalID) || q.RootID != "" && !validUUID(q.RootID) || len(q.Cursor) > 2000 {
		return StructurePage{}, &Error{Code: CodeInvalidRequest}
	}
	st, err := s.structureStore()
	if err != nil {
		return StructurePage{}, err
	}
	return st.ReadStructure(ctx, q)
}
func (s *Service) ReadConcept(ctx context.Context, id, revision string) (StructureNode, error) {
	if !validUUID(id) || revision != "" && !validUUID(revision) {
		return StructureNode{}, &Error{Code: CodeInvalidRequest}
	}
	st, err := s.structureStore()
	if err != nil {
		return StructureNode{}, err
	}
	return st.ReadConcept(ctx, id, revision)
}
func (s *Service) CreateStructureProposal(ctx context.Context, c StructureCommand) (StructureProposal, error) {
	if err := c.Validate(); err != nil {
		return StructureProposal{}, err
	}
	st, err := s.structureStore()
	if err != nil {
		return StructureProposal{}, err
	}
	return st.CreateStructureProposal(ctx, c)
}
func (s *Service) ReadStructureProposal(ctx context.Context, id string) (StructureProposal, error) {
	if !validUUID(id) {
		return StructureProposal{}, &Error{Code: CodeInvalidRequest}
	}
	st, err := s.structureStore()
	if err != nil {
		return StructureProposal{}, err
	}
	return st.ReadStructureProposal(ctx, id)
}
func (s *Service) ListStructureProposals(ctx context.Context, cursor string, limit int) (StructureProposalPage, error) {
	if limit == 0 {
		limit = 20
	}
	if limit < 1 || limit > 100 || cursor != "" && !validUUID(cursor) {
		return StructureProposalPage{}, &Error{Code: CodeInvalidRequest}
	}
	st, err := s.structureStore()
	if err != nil {
		return StructureProposalPage{}, err
	}
	return st.ListStructureProposals(ctx, cursor, limit)
}
func (s *Service) DecideStructureProposal(ctx context.Context, id string, c StructureDecision) (StructureProposal, error) {
	if !validUUID(id) || !validUUID(c.OperationID) || !validUUID(c.ActorDeviceID) || len(c.Hash) != 64 || !structureText(c.Reason, 2000, true) || (c.Decision != "approve" && c.Decision != "reject") {
		return StructureProposal{}, &Error{Code: CodeInvalidRequest}
	}
	st, err := s.structureStore()
	if err != nil {
		return StructureProposal{}, err
	}
	return st.DecideStructureProposal(ctx, id, c)
}

File: clients/web/src/api/structure.ts
import { z } from 'zod'
import { learningClient, unwrap } from './client'
import type { Session } from './runtime'

const id = z.uuid(), version = z.number().int().nonnegative().max(Number.MAX_SAFE_INTEGER)
export const sourceStates = { candidate: '候选', included: '已纳入', unverified: '待核实', conflict: '冲突', superseded: '已替代' }
export const learningStates = { unseen: '未接触', needs_practice: '待练习', insufficient_evidence: '证据不足', evidenced: '有证据' }
export const relationNames = { prerequisite: '前置', related: '相关', contrast: '对比', part_of: '组成' }
export const proposalStates = { open: '待审阅', applied: '已应用', rejected: '已拒绝', stale: '已过期' }
export const conceptSource = z.object({ collection_id: id, revision_id: id, document_id: id, node_id: id, quote: z.string() })
export const conceptContent = z.object({ description: z.string(), source_status: z.enum(['candidate', 'included', 'unverified', 'conflict', 'superseded']), suggested: z.boolean(), sources: z.array(conceptSource), claims: z.array(z.object({ text: z.string(), conditions: z.string(), sources: z.array(version), gap: z.string() })), relations: z.array(z.object({ target_id: id, kind: z.enum(['prerequisite', 'related', 'contrast', 'part_of']), suggested: z.boolean(), sources: z.array(version) })), replaced_by: z.array(id) })
export const structureNode = z.object({ concept_id: id, revision_id: id, semantic_key: z.string(), name: z.string(), support: z.array(z.object({ source_id: id, revision_id: id, fragment_id: id, quote: z.string() })), goal_id: z.string(), content: conceptContent, learning_state: z.enum(['unseen', 'needs_practice', 'insufficient_evidence', 'evidenced']) })
export const structurePage = z.object({ version, generation: version, items: z.array(structureNode), edges: z.array(conceptContent.shape.relations.element.extend({ source_id: id })), next_cursor: z.string(), partial: z.boolean(), notice: z.string() })
export const structureCapabilities = z.object({ protocol_version: z.literal(1), available: z.boolean(), can_propose: z.boolean(), can_decide: z.boolean(), max_nodes: version, max_edits: version })
const impact = z.object({ goal_ids: z.array(id), contexts: version, activities: version, contents: version, evidence: version, fingerprint: z.string() })
export const structureProposal = z.object({ id, operation_id: id, status: z.enum(['open', 'applied', 'rejected', 'stale']), kind: z.enum(['edit', 'merge', 'split', 'compensate']), reason: z.string(), base_version: version, generation: version, hash: z.string(), before: z.array(structureNode), after: z.array(structureNode), impact, current_impact: impact, compensates: z.string(), applied_version: version, created_at: z.string(), decision_reason: z.string(), replayed: z.boolean() })
export const structureProposals = z.object({ items: z.array(structureProposal), next_cursor: z.string() })
export type StructureNode = z.infer<typeof structureNode>
export type ConceptContent = z.infer<typeof conceptContent>
export type StructureProposal = z.infer<typeof structureProposal>
export type StructureEdit = Pick<StructureNode, 'concept_id' | 'goal_id' | 'name' | 'content'>
export type StructureCommand = { operation_id: string; base_version: number; generation: number; kind: 'edit' | 'merge' | 'split' | 'compensate'; reason: string; edits: StructureEdit[]; compensates?: string }
export type StructureDecision = { operation_id: string; hash: string; decision: 'approve' | 'reject'; reason: string }
type Response<T> = { responses: { 200: { content: { 'application/json': T } } } }
type Post<B, T> = Response<T> & { requestBody: { content: { 'application/json': B } } }
type Path<P> = { parameters: { path: P } }
export interface StructurePaths {
  '/v1/knowledge/structure/capabilities': { get: Response<z.infer<typeof structureCapabilities>> }
  '/v1/knowledge/structure': { get: Response<z.infer<typeof structurePage>> & { parameters: { query?: { goal_id?: string; root_id?: string; search?: string; cursor?: string; limit?: number } } } }
  '/v1/knowledge/structure/concepts/{conceptID}': { get: Response<StructureNode> & { parameters: { path: { conceptID: string }; query?: { revision_id?: string } } } }
  '/v1/knowledge/structure/proposals': { get: Response<z.infer<typeof structureProposals>> & { parameters: { query?: { cursor?: string; limit?: number } } }; post: Post<StructureCommand, StructureProposal> }
  '/v1/knowledge/structure/proposals/{proposalID}': { get: Response<StructureProposal> & Path<{ proposalID: string }> }
  '/v1/knowledge/structure/proposals/{proposalID}/decisions': { post: Post<StructureDecision, StructureProposal> & Path<{ proposalID: string }> }
}
export function structureAPI(session: Session, space: string) {
  const client = () => learningClient(session, space)
  return {
    capabilities: () => unwrap(client().GET('/v1/knowledge/structure/capabilities'), structureCapabilities),
    list: (query: { goal_id?: string; root_id?: string; search?: string; cursor?: string; limit?: number }) => unwrap(client().GET('/v1/knowledge/structure', { params: { query } }), structurePage),
    node: (conceptID: string, revision_id?: string) => unwrap(client().GET('/v1/knowledge/structure/concepts/{conceptID}', { params: { path: { conceptID }, query: { revision_id } } }), structureNode),
    proposals: (cursor = '') => unwrap(client().GET('/v1/knowledge/structure/proposals', { params: { query: { cursor, limit: 20 } } }), structureProposals),
    proposal: (proposalID: string) => unwrap(client().GET('/v1/knowledge/structure/proposals/{proposalID}', { params: { path: { proposalID } } }), structureProposal),
    create: (body: StructureCommand) => unwrap(client().POST('/v1/knowledge/structure/proposals', { body }), structureProposal),
    decide: (proposalID: string, body: StructureDecision) => unwrap(client().POST('/v1/knowledge/structure/proposals/{proposalID}/decisions', { params: { path: { proposalID } }, body }), structureProposal),
  }
}

export const emptyConcept = (): ConceptContent => ({ description: '', source_status: 'candidate', suggested: false, sources: [], claims: [], relations: [], replaced_by: [] })
